railties/lib/rails/commands/rake/rake_command.rb

```diff
@@ -18,7 +18,9 @@ def perform(task, args, config)
           Rake.with_application do |rake|
             rake.init("rails", [task, *args])
             rake.load_rakefile
+
             if unrecognized_task = rake.top_level_tasks.find { |task| !rake.lookup(task[/[^\[]+/]) }
+              @rake_tasks = rake.tasks
```

Comment:
Two different paths to set a memoized value makes me nervous, especially given the sequence of steps involved.

Any way we can unify them?

Reply:
The primary difference between the two paths is the way the Rake tasks are loaded.  This difference dates back to the original implementation:

https://github.com/rails/rails/blob/6813edc7d926965e5644cd8befaf229a35b9d8ca/railties/lib/rails/commands/rake/rake_command.rb#L11-L19

https://github.com/rails/rails/blob/6813edc7d926965e5644cd8befaf229a35b9d8ca/railties/lib/rails/commands/rake/rake_command.rb#L24-L36

I'm not 100% sure why it was written that way.  My guess is the purpose was to exclude tasks defined in the application's `Rakefile` from being listed in the help output.  But that would not exclude tasks defined in `lib/tasks/*.rake`.

Other than that, the two paths end up being equivalent for the default generated `Rakefile`.

* `RakeCommand::perform` calls `Rake::Application#load_rakefile`, which then loads the app and calls `Rails.application.load_tasks`:

  https://github.com/rails/rails/blob/ba19dbc49956a73f417abd68c7a5f33e302eacd3/railties/lib/rails/generators/rails/app/templates/Rakefile.tt#L4-L6

* `RakeCommand::rake_tasks` loads the app:

  https://github.com/rails/rails/blob/ba19dbc49956a73f417abd68c7a5f33e302eacd3/railties/lib/rails/command/actions.rb#L13-L16

  and calls `Rails.application.load_tasks`:

  https://github.com/rails/rails/blob/ba19dbc49956a73f417abd68c7a5f33e302eacd3/railties/lib/rails/command/actions.rb#L42-L44

  * And for an engine, it just calls `Rake::Application#load_rakefile` directly:

    https://github.com/rails/rails/blob/ba19dbc49956a73f417abd68c7a5f33e302eacd3/railties/lib/rails/command/actions.rb#L31-L34

That being said, we can exclude tasks from the help by using `Rake::Task#locations` instead.  Doing so also allows us to filter out tasks defined by the application in `lib/tasks/*.rake`.

I also think it makes sense to load tasks in the same way in both `::perform` and `::rake_tasks`.  That way `::rake_tasks` only lists tasks which can actually be run by `::perform` (in case those lists diverge somehow).

So I've added a couple of commits to make those changes.  However, I've kept `@rake_tasks` in `::perform` because I think it fits conceptually.  Basically, `::perform` is preserving its tasks before the `Rake::Application` instance is thrown away.  Then, if any tasks have been preserved, `::rake_tasks` returns them; otherwise, `::rake_tasks` loads tasks in the same way that `::perform` would and returns those.